### ISSUE TITLE
Conflict with projections.nvim

### ISSUE THREAD

**Issue by nyngwang** [nvim-neo-tree/neo-tree.nvim#651]
### Problem Description

I'm using neo-tree with projections.nvim. As shown in the demo provided below, I got an error like this when I called `:Neotree focus`:

```
debounce  filesystem_navigate  error:  ...im/site/pack/packer/start/nui.nvim/lua/nui/tree/init.lua:254: Expected Lua number
```

<img src="https://user-images.githubusercontent.com/24765272/206854965-ef147c7e-7560-4bd2-983d-1451affc2450.png" width="600">

### Demo

This is the full demo on how to reproduce the error:

https://user-images.githubusercontent.com/24765272/206855146-c1d0e29d-5fe8-4723-b1c6-45dc203d91e5.mov

### Config for reproducing

The config of projections.nvim:

```lua
use {
  'gnikdroy/projections.nvim',
  config = function ()
    require('projections').setup {
      -- workspaces = require('project.projects'),
      patterns = {
        'init.lua',
        '.git',
      },
      store_hooks = {
        pre = function ()
          vim.cmd('tabd Neotree close')
          vim.cmd('tabn')
        end,
        post = function () end,
      },
      restore_hooks = {
        pre = function () end,
        post = function () end,
      },
    }
    -- auto-store on any leave/quit.
    vim.api.nvim_create_autocmd({ 'VimLeavePre', 'DirChangedPre' }, {
      group = 'session.lua',
      pattern = '*',
      callback = function ()
        require('projections.session').store(vim.loop.cwd())
      end
    })
    -- auto-restore on start excluding `gm` and `nvim ...`.
    vim.api.nvim_create_autocmd({ 'VimEnter', 'DirChanged' }, {
      group = 'session.lua',
      pattern = '*',
      callback = function ()
        if vim.fn.argc() == 0 then
          require('projections.switcher').switch(vim.loop.cwd())
        end
      end
    })
  end
}
```

**Comment by nyngwang** [comment]
If I run `:Neotree close` before I run `:Neotree focus`, then there is no error.

**Comment by nyngwang** [comment]
With the help of #639, I can confirm that at `0:06` of the demo, neo-tree seems to be internally considered open, while as you can see, there is no neo-tree buffer there starting from `0:06`.

**Comment by cseickel** [comment]
That is actually the ideal way to check if neo-tree is open. Does having that function alleviate the problem?

**Comment by nyngwang** [comment]
To @cseickel:

> [...] Does having that function alleviate the problem?

Yes. But this doesn't solve the root cause of this problem: the `debounce  filesystem_navigate  error` is still there. I think this problem is highly related to #633, to quote your words:

> If the cursor is in the neo-tree window when you call a fzf api, it will probably try to open whatever file you selected within the Neo-tree window. Neo-tree will then detect that it's window was stolen, and move that intruder to a new window and then reclaim the window that Neo-tree originally was in. [...]

**Comment by cseickel** [comment]
On the subject of whether neo-tree is open, my first code snippet was just wrong so don't go by that. Your function should be accurate though. You also need to consider that just because you can't see it being open, does not mean it is not open at the instant that the code runs. It can easily open and close so fast you can't see it happening. 

There is certainly a bug somewhere. My impression is that neo-tree is being closed in the middle of being shown, at a place in the code where I did not anticipate the window going away.

What exactly is supposed to be happening here? I'm not familiar with projections.nvim.

**Comment by nyngwang** [comment]
To @cseickel: Sorry for my late reply, I fell asleep.

> [...] It can easily open and close so fast you can't see it happening.
> [...] My impression is that neo-tree is being closed in the middle of being shown, at a place in the code where I did not anticipate the window going away.

Understood. Put in simple words: I think neo-tree's internal logic conflicts with the `store_hooks`,`restore_hooks` of projections.nvim. I will explain it below.

> What exactly is supposed to be happening here? I'm not familiar with projections.nvim.

I have created the following `autocmd`. With it, the callback will store the current project upon `DirChangedPre`. I use neo-tree's `set_root` to navigate between folders, and this triggers `DirChangedPre`. Once `store` in the callback is called, the callback specified in `store_hooks.pre` will be called. Since I put `tabd Neotree close` in it, I expect **no** neo-tree window will be stored in my project session (since it's problematic when restoring). But this didn't happen yesterday, these are my conclusion:

- Due to the design you said in #633, **neo-tree is hard to close**: The `store_hooks.pre` failed to close all neo-tree buffers yesterday, there were some "strange, empty, `filetype=neo-tree` windows" hanging there in my project after restoring. The author of projections.nvim claimed that the problem was fixed as there are no more such empty windows after the fix.
- Now there is no "neo-tree window", but it doesn't get close: this is hilarious, to be honest. While the author of projections.nvim tried his best to fix the `store_hooks`, `restore_hooks`. This actually makes this problem hard to solve since I don't even know how to debug an invisible window.

```lua
vim.api.nvim_create_autocmd({ 'VimLeavePre', 'DirChangedPre' }, {
  pattern = '*',
  callback = function ()
    require('projections.session').store(vim.loop.cwd())
  end
})
require('projections').setup {
  store_hooks = {
    pre = function ()
      vim.cmd('tabd Neotree close')
      vim.cmd('tabn')
    end,
    post = function () end,
  },
  restore_hooks = {
    pre = function () end,
    post = function ()
      vim.cmd('tabd Neotree close')
      vim.cmd('tabn')
    end,
  },
}
```

**Comment by cseickel** [comment]
> I have created a `autocmd` as shown below. With it, the callback will store the current project upon `DirChangedPre`. I use neo-tree's `set_root` to navigate between folders, and this triggers `DirChangedPre`. Once `store` in the callback is called, the callback specified in `store_hooks.pre` will be called. Since I put `tabd Neotree close` in it,

That sounds like a bad idea. If you are using Neo-tree to change directories, and you have an autocmd to close neo-tree when you change directories, you are causing the bug by intentionally closing neo-tree when it is in the middle of refreshing. If you don't want neo-tree to be included in the session, then ideally you should either:

- Configure projections.nvim to exclude Neo-tree from the session, usually session plugins have a config to exclude windows by filetype.
- If it doesn't have that option, use a session plugin that does.
- Or submit a PR to projections.nvim that adds this functionality.

Definitely do not intentionally close Neo-tree during a directory change while `bind_to_cwd` is enabled.

It is entirely possible that this problem does not happen for someone else but does for you based on what other plugins and autocmds you have configured. It is essentially a timing issue where you could get lucky and have the various conflicting actions work out in the correct order.

All that being said, I can and did add more protections to deal with a window being closed in the middle of a render. I have tried to do that in PR #654, which was just merged to `main`. Try it out and let me know if it helps.

**Comment by nyngwang** [comment]
@cseickel Really happy to see your detailed suggestion and already-done PR :) But to be honest, I've removed the plugin (the cost of communication with the author is just too high for both of us, around 5-10 hours per issue. I felt sorry about bombard him with my messy wordy issues). Instead, I've created one tailored for myself and it works very well with neo-tree.lua.

> All that being said, I can and did add more protections to deal with a window being closed in the middle of a render. I have tried to do that in PR https://github.com/nvim-neo-tree/neo-tree.nvim/pull/654, which was just merged to main. Try it out and let me know if it helps.

You did. It works well with my plugin.

> If you don't want neo-tree to be included in the session, then ideally you should either: [...]

I've added an example to my [README.md](https://github.com/nyngwang/suave.lua#setup). Maybe I should put emphasis on how good neo-tree is (while the freedom it provides does make it intimidating for beginners) and recommend people to use these two together. Anyway, really sorry that I kinda made here messy and I have no energy for resolving this problem since I no longer use it.

It's OK to help others though, once some people encounter the same or similar problem I will try my best to help. Closing... Really thank you for your careful helpings.
